Angular 16 esbuild: inline JavaScript is not enabled of less

### Command

serve

### Description

Hope to keep `javascriptEnabled` enabled in esbuild, currently ng-zorro-antd still needs support for inline scripts.

Config differences between [esbuild](https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/less-language.ts#L90) and [webpack](https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/tools/webpack/configs/styles.ts#L272).

A simple step to reproduce:

```
# Angular 16.1.x
ng new demo
ng add ng-zorro-antd
ng s --force-esbuild
```

Error:

```
✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/src/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^


✘ [ERROR] Inline JavaScript is not enabled. Is it set in your options? [plugin angular-less]

    node_modules/ng-zorro-antd/style/color/bezierEasing.less:110:0:
      110 │ .bezierEasingMixin();
          ╵ ^

  The plugin "angular-less" was triggered by this import

    angular:styles/global:styles:1:8:
      1 │ @import 'src/styles.less';
        ╵         ~~~~~~~~~~~~~~~~~


  ➜  Local:   http://127.0.0.1:4200/
Watch mode enabled. Watching for file changes...
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/bundle-options.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/bundle-options.ts
@@ -64,7 +64,7 @@ export function createStylesheetBundleOptions(
     preserveSymlinks: options.preserveSymlinks,
     external: options.externalDependencies,
     publicPath: options.publicPath,
-    conditions: ['style', 'sass'],
+    conditions: ['style', 'sass', 'less'],
     mainFields: ['style', 'sass'],
     plugins: [
       pluginFactory.create(SassStylesheetLanguage),

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/less-language.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/less-language.ts
@@ -6,7 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import type { OnLoadResult, PluginBuild } from 'esbuild';
+import type { Location, OnLoadResult, PluginBuild } from 'esbuild';
 import { readFile } from 'node:fs/promises';
 import { StylesheetLanguage, StylesheetPluginOptions } from './stylesheet-plugin-factory';
 
@@ -32,7 +32,13 @@ export const LessStylesheetLanguage = Object.freeze<StylesheetLanguage>({
   componentFilter: /^less;/,
   fileFilter: /\.less$/,
   process(data, file, _, options, build) {
-    return compileString(data, file, options, build.resolve.bind(build));
+    return compileString(
+      data,
+      file,
+      options,
+      build.resolve.bind(build),
+      /* unsafeInlineJavaScript */ false,
+    );
   },
 });
 
@@ -41,6 +47,7 @@ async function compileString(
   filename: string,
   options: StylesheetPluginOptions,
   resolver: PluginBuild['resolve'],
+  unsafeInlineJavaScript: boolean,
 ): Promise<OnLoadResult> {
   const less = (lessPreprocessor ??= (await import('less')).default);
 
@@ -92,6 +99,7 @@ async function compileString(
       paths: options.includePaths,
       plugins: [resolverPlugin],
       rewriteUrls: 'all',
+      javascriptEnabled: unsafeInlineJavaScript,
       sourceMap: options.sourcemap
         ? {
             sourceMapFileInline: true,
@@ -107,17 +115,40 @@ async function compileString(
     };
   } catch (error) {
     if (isLessException(error)) {
+      // Retry with a warning for less files requiring the deprecated inline JavaScript option
+      if (error.message.includes('Inline JavaScript is not enabled.')) {
+        const withJsResult = await compileString(
+          data,
+          filename,
+          options,
+          resolver,
+          /* unsafeInlineJavaScript */ true,
+        );
+        withJsResult.warnings = [
+          {
+            text: 'Deprecated inline execution of JavaScript has been enabled ("javascriptEnabled")',
+            location: convertExceptionLocation(error),
+            notes: [
+              {
+                location: null,
+                text: 'JavaScript found within less stylesheets may be executed at build time. [https://lesscss.org/usage/#less-options]',
+              },
+              {
+                location: null,
+                text: 'Support for "javascriptEnabled" may be removed from the Angular CLI starting with Angular v19.',
+              },
+            ],
+          },
+        ];
+
+        return withJsResult;
+      }
+
       return {
         errors: [
           {
             text: error.message,
-            location: {
-              file: error.filename,
-              line: error.line,
-              column: error.column,
-              // Middle element represents the line containing the error
-              lineText: error.extract && error.extract[Math.trunc(error.extract.length / 2)],
-            },
+            location: convertExceptionLocation(error),
           },
         ],
         loader: 'css',
@@ -126,4 +157,14 @@ async function compileString(
 
     throw error;
   }
+}
+
+function convertExceptionLocation(exception: LessException): Partial<Location> {
+  return {
+    file: exception.filename,
+    line: exception.line,
+    column: exception.column,
+    // Middle element represents the line containing the exception
+    lineText: exception.extract && exception.extract[Math.trunc(exception.extract.length / 2)],
+  };
 }

